lms/services/canvas_api.py

```diff
@@ -91,10 +94,25 @@ def public_url(self, file_id):
 
         :rtype: str
         """
-        return self._helper.validated_response(
-            self._helper.public_url_request(self._oauth2_token.access_token, file_id),
+        oauth2_token = self._oauth2_token
+
+        return self.send_with_refresh_and_retry(
+            self._helper.public_url_request(oauth2_token.access_token, file_id),
             CanvasPublicURLResponseSchema,
-        ).parsed_params["public_url"]
+            oauth2_token.refresh_token,
+        )["public_url"]
+
+    def send_with_refresh_and_retry(self, request, schema, refresh_token):
+        try:
+            return self._helper.validated_response(request, schema).parsed_params
+        except CanvasAPIAccessTokenError:
+            if not refresh_token:
+                raise
+
+            new_access_token = self.get_refreshed_token(refresh_token)
```

Comment:
`get_refreshed_token()` saves the new access token to the DB for next time, as well as returning it for right now.